### DISCUSSION TITLE
how do I make a TextEdit in a Window that a user can resize but does ot resize itself.

### DISCUSSION THREAD

**Discussion by pm100** [emilk/egui#4799] [Q&A] [answered]
I am making a console window. At the moment its a Window>ScrollArea>TextEdit. I have most of it working but I cant see how to prevent the TextEdit forcing the Window to grow vertically as the user types, or the system responds. The scroll bar does not kick in until the window has grown a lot.

**Comment by emilk** [comment]
Sounds like you want `ScrollArea::max_height`: https://docs.rs/egui/latest/egui/containers/scroll_area/struct.ScrollArea.html#method.max_height

**Comment by YgorSouza** [comment, marked as answer]
It seems to work fine if I add a `ScrollArea` to this example:

https://github.com/emilk/egui/blob/cd1e4c573a963d25db78c12daf998dd059a2f5c1/crates/egui_demo_lib/src/demo/tests/window_resize_test.rs#L74-L82

```rs
        Window::new("↔ resizable with TextEdit")
            .open(open)
            .vscroll(false)
            .resizable(true)
            .default_height(300.0)
            .show(ctx, |ui| {
                ui.label("Shows how you can fill an area with a widget.");
                ScrollArea::vertical().show(ui, |ui| {
                    ui.add_sized(ui.available_size(), TextEdit::multiline(&mut self.text));
                });
            });
```

https://github.com/emilk/egui/assets/43298013/4e16c1e1-3746-42db-ba0f-ed298f994da3

**Reply by pm100** [reply]
TY, trying to morph my code into your code, mine is phrased slightly differently so I can get the cursor state of the textedit. I think its the add_sized that does the trick (I changed everything else to match and it made no difference to my apps behavior) but I cannot work out how to add it here

```
// window already done, its ui is whats passed in here
        egui::ScrollArea::vertical().show(ui, |ui| {
            let widget = egui::TextEdit::multiline(&mut self.text)
                .font(egui::TextStyle::Monospace) // for cursor height
                .code_editor()
                // .desired_rows(10)
                .lock_focus(true)
                .desired_width(f32::INFINITY)
                .id(Id::new("console_text"))
                .show(ui);
            let mut new_cursor = None;
            self.last_cursor = usize::MAX;
            if let Some(cursor) = widget.state.cursor.char_range() {
                ..... cursor managment code
```

the different combinations of ui.add, ui.show and what they each return has me befuddled

**Reply by YgorSouza** [reply]
I think you can just put all that inside the `add_sized`:

```rs
ScrollArea::vertical().show(ui, |ui| {
    ui.add_sized(ui.available_size(), |ui: &mut Ui| {
        let text_edit = TextEdit::multiline(&mut self.text)
            .code_editor()
            .lock_focus(true)
            .id(Id::new("console_text"));
        let output = text_edit.show(ui);
        if let Some(cursor) = output.state.cursor.char_range() {
            //... handle that
        }
        output.response
    });
});
```

**Reply by pm100** [reply]
woohoo - thanks.

I really have to say that the three different ways of dealing with widgets
- ui.add...
- ui.button.....
- widget.show(ui)

are very confusing. Especially the inversion of the last one, widget does tuff that uses Ui object as opposed to Ui does stuff that uses widgets. Also understanding what these different return types mean. I promise I am not dumb but its hard to follow

**Reply by YgorSouza** [reply]
The `Ui` methods like `button`, `label`, `horizontal` etc are to save time when using the most common widgets and layouts with default settings, since a lot of people use egui as a debugging tool and don't want to spend too much time styling it. Then when you want to create your own widgets, you implement the `Widget` trait, using the `show` method to specify how it show draw itself and interact with the user input. Then egui has a blanket implementation of `Widget` for functions, so you can just declare a function or even a closure for a simple one-off widget, instead of having to create a whole new type. So you have `Ui::add` to allow you to put these function-widgets in a UI. It also allows you to use your regular widgets without having to import the `Widget` trait everywhere to get access to the `show` method.

https://github.com/emilk/egui/blob/f9476e0a00a6622d2f01b56964bba9e97fd065ed/crates/egui/src/widgets/mod.rs#L72-L95

But creating a type that implements `Widget` instead of just a function is useful if you want to add options to it, so you can create builder methods that modify these options before passing it to the UI, like the default widgets that are declared in egui.
